### PR TITLE
Add ability to perform automatic tidy operations

### PR DESCRIPTION
**Pull request by cipherboy**
This enables the PKI secrets engine to allow tidy to be started
periodically by the engine itself, avoiding the need for interaction.
This operation is disabled by default (to avoid load on clusters which
don't need tidy to be run) but can be enabled.

In particular, a default tidy configuration is written (via
`/config/auto-tidy`) which mirrors the options passed to `/tidy`. Two
additional parameters, enabled and interval, are accepted, allowing
auto-tidy to be enabled or disabled and controlling the interval
(between successful tidy runs) to attempt auto-tidy.

Notably, a manual execution of tidy will delay additional auto-tidy
operations. Status is reported via the existing `/tidy-status` endpoint.

`Signed-off-by: Alexander Scheel <alex.scheel@hashicorp.com>`

---

 - [x] Docs
 - [x] Tests
 - [x] Changelog

### CODE REVIEW

File: builtin/logical/pki/path_tidy_test.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+	"github.com/hashicorp/vault/api"
+	vaulthttp "github.com/hashicorp/vault/http"
+	"github.com/hashicorp/vault/sdk/logical"
+	"github.com/hashicorp/vault/vault"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestAutoTidy(t *testing.T) {
```
Comment (cipherboy) [inline builtin/logical/pki/path_tidy_test.go:15]: Test takes 17s as it is fwiw.

File: builtin/logical/pki/path_tidy.go [hunk 2]
```diff
@@ -411,6 +429,65 @@ func (b *backend) pathTidyStatusRead(_ context.Context, _ *logical.Request, _ *f
...
+	}
+
+	if safetyBufferRaw, ok := d.GetOk("safety_buffer"); ok {
+		config.SafetyBuffer = time.Duration(safetyBufferRaw.(int)) * time.Second
+		if config.SafetyBuffer < 1*time.Second {
+			return logical.ErrorResponse(fmt.Sprintf("given safety_buffer must be greater than zero seconds; got: %v", safetyBufferRaw)), nil
+		}
+	}
+
+	return nil, sc.writeAutoTidyConfig(config)
```
Comment (stevendpclark) [inline builtin/logical/pki/path_tidy.go:488]: If an end-user has enabled auto-tidy but has not enabled any of the flags to actually do the cleaning should we not warn/error out to them here like we do on the scheduling api?

File: builtin/logical/pki/path_tidy.go [hunk 3]
```diff
@@ -411,6 +429,65 @@ func (b *backend) pathTidyStatusRead(_ context.Context, _ *logical.Request, _ *f
...
+	sc := b.makeStorageContext(ctx, req.Storage)
+	config, err := sc.getAutoTidyConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"enabled":                               config.Enabled,
+			"interval":                              int(config.Interval / time.Second),
```
[the remark this answers is not in the document]
Comment (cipherboy) [reply builtin/logical/pki/path_tidy.go:442]: `_duration` I believe?

[the remark this answers is not in the document]
Comment (cipherboy) [reply builtin/logical/pki/path_tidy.go:442]: I could do `interval_duration` if we want, and make that consistent across the input/output, but I kinda liked the idea of having output parameters be the same as input parameters; that way they can easily backup/restore &c.

File: builtin/logical/pki/path_tidy.go (hunk 2 above)
Comment (cipherboy) [reply builtin/logical/pki/path_tidy.go:488]: I made it an error :-)

**Comment by cipherboy** [comment]
@stevendpclark updated and bumped the timeout on this test... We'll see if that helps? :-)

File: website/content/api-docs/secret/pki.mdx [hunk 4]
```diff
@@ -3204,6 +3208,66 @@ $ curl \
     http://127.0.0.1:8200/v1/pki/tidy
 ```
 
+### Configure Automatic Tidy
```
Comment (stevendpclark) [inline website/content/api-docs/secret/pki.mdx:3211]: Should we add a small blurb that the existing cancellation of tidy operations also work with automatically started tidy ops?

**Review by stevendpclark — APPROVED** [review]
Looks good to me, one nit which I feel pretty ambivalent about either way on the docs.

File: website/content/api-docs/secret/pki.mdx (hunk 4 above)
Comment (cipherboy) [reply website/content/api-docs/secret/pki.mdx:3211]: Tidy cancellation is not yet implemented :-)

**Review by stevendpclark — APPROVED** [review]

**Comment by cipherboy** [comment]
Thanks! Finally got all the data races figured out! ty!

File: builtin/logical/pki/path_tidy.go [hunk 5]
```diff
@@ -88,6 +63,36 @@ func pathTidyStatus(b *backend) *framework.Path {
...
+		Pattern: "config/auto-tidy",
+		Fields: addTidyFields(map[string]*framework.FieldSchema{
+			"enabled": {
+				Type:        framework.TypeBool,
+				Description: `Set to true to enable automatic tidy operations.`,
+			},
+			"interval_duration": {
+				Type:        framework.TypeDurationSecond,
+				Description: `Interval at which to run an auto-tidy operation. This is the time between tidy invocations (after one finishes to the start of the next). Running a manual tidy will reset this duration.`,
+				Default:     43200, // 32h, but TypeDurationSecond currently requires the default to be an int.
```
Comment (beornf) [inline builtin/logical/pki/path_tidy.go:77]: 43200s = 12h. Is 32h default a typo?

Comment (cipherboy) [reply builtin/logical/pki/path_tidy.go:77]: Ah, good eyes. Yes, a typo. Note though that auto-tidy isn't enabled by default.

**Comment by abh** [comment]
Very nice; I am looking forward to trying it out!

**Pull request merged**